### PR TITLE
Multiblock machine for Fluid Solidifier

### PR DESCRIPTION
**Pull request by OmdaCZ**
Continuing the project to deprecate Large Processing Factory, this PR adds a multiblock for the Fluid Solidifier.
Planned scaling:
Speed boost dependent on fluid used to cool down the molten metals faster
Parallels scale from better material molds and the width of the multi adding more space for those molds.
Maximal width relying on tiered casings, hopefully at least 3 tiers from Early IV after Alloy Blast Smelter to perhaps UMV.

Current situation of development is stuck on structural check. (I am open to help from more professional people)

Multiblock structure smallest and largest (unfinished) to 1/8/2024:

### CODE REVIEW

**Comment by Nxer** [comment]
Crazy structure with crazy color.

**Comment by EnderProyects** [comment]
Wasn't there an unwritten rule not to make new multiblocks with a cubic shape, but rather to make them irregular?

**Comment by OmdaCZ** [comment]
The shape of the structure can easily be fixed to be more aesthetically pleasing to more people.

**Comment by Nxer** [comment]
Can machine be built in two different length in both sides?

**Comment by OmdaCZ** [comment]
No, at least hologram projector won't build it unevenly.

**Comment by Dream-Master** [comment]
@OmdaCZ what are the status of this pr ?

**Comment by OmdaCZ** [comment]
> @OmdaCZ what are the status of this pr ?

Currently implementing tiered casings for maximal width.

**Comment by OneEyeMaker** [comment]
> > @OmdaCZ what are the status of this pr ?
> 
> Currently implementing tiered casings for maximal width.

Why should every multiblock have its own casings? Why not use existing ones?

**Comment by OmdaCZ** [comment]
it does use its own casings for t1, but borrows some for t2 and t3

**Comment by Volence** [comment]
Would it be possible/wanted to have this also be able to have the ability to do the recipes that the Tool Casting Machines can handle? They work in a very similar way and it would be nice to have a multi that handles their recipes as well instead of several single blocks

**Comment by OmdaCZ** [comment]
Current status:
Parallels dont work as they should (@FourIsTheNumber should help fix that)
Structures of min tier and max tier (and size):

current max parallels should be around 60, speed bonus hard fixed 100%.
I'm open to balancing changes to speed/parallel/eu discount.

**Comment by FireWarrior2** [comment]
Tbh, max size structure looks very strange and even ugly. 
Maybe instead of such structure extension just make it use some expensive blocks (like PCB)?..

**Comment by joshaup1999** [comment]
60 max parallel is way to small. I know i will more than likely still go with a PA array over this multi if it stays with the current structure and parallel amount

**Comment by ogdae** [comment]
As per Jo's comment, 60 parallel cap is way way too small. PA would still be way more viable over this. Additionally, the single expanding structure sideways just doesn't really fit well, in my opinion, especially against the slew of machines being released with more aesthetic and intricate additions to their tiers. I can't really see the casing art but from a distance it looks okay, but including casting tables/basins and hoppers in the structure check seems very odd.

That aside, I quite like a couple of the proposed mechanics, specifically cooling requirements and mold-tier, however we currently only have access to steel and then white dwarf matter molds, so not sure how mold-tier bonuses would be implemented well.

A solid idea but the structure certainly needs work, and maybe a reduction in mechanics to one really important one or two that work in tandem.

**Comment by FourIsTheNumber** [comment]
> 60 max parallel is way to small. I know i will more than likely still go with a PA array over this multi if it stays with the current structure and parallel amount

PA will not exist, so this is an irrelevant comparison point. The correct comparison is LPF, which gets 2 parallels per tier :P 
This multi is certainly far far too large to have those stats, but that is an issue with the structure, not the parallel count.

**Comment by Ruling-0** [comment]
I think it's a little awkward to have such a large multi for only 60 parallels - consider how many LPFs you could fit in that space, which are x3.5 speed (I'm not sure what the speed stats of this multi are). On checking, without even wallsharing them you can fit 39 to 40 LPFs in that footprint lengthwise, which seems to be a lot more throughput per space, while being both easier to integrate into non-lawnbase builds and easier to use in lines that just need a bit of fluid solidifying compared to the min size of this multi.

The cooling mechanic sounds cool, and the min-size footprint is workable, but the max-size footprint seems like the width-increasing-parallel mechanic was taken too far, esp. for a multi that comes in IV. This stands out _a lot_ from the other IV multis, and trying to make it scale from IV to UMV in such an integral mechanic makes it harder to potentially narrow that scaling later and introduce a late-game solidifer, compared to just making high-tier fluids stupid fast or having higher numbers to compensate for that late-game solidifer not existing yet.

**Comment by Dream-Master** [comment]
@OmdaCZ @FourIsTheNumber but this is not for 2.7.0? or is this done and can be tested with the other stuff in a new RC version soon ?

**Comment by FourIsTheNumber** [comment]
> @OmdaCZ @FourIsTheNumber but this is not for 2.7.0? or is this done and can be tested with the other stuff in a new RC version soon ?

IMO this (or _some_ solidifier multi) needs to be for 2.7, as it is the last machine required to replace LPF. I don't want to have to wait another patch cycle for literally one machine

**Comment by Dream-Master** [comment]
ok i lift the freeze but new stuff will be definitely freeze

**Comment by Dream-Master** [comment]
@OmdaCZ @FourIsTheNumber so pr is done? i asl @NotAPenguin0 for a review

**Comment by boubou19** [comment]
the big renaming is being worked on, so this must not be merged until it is done, or you'll cause massive conflicts to Penguin's PR.

**Comment by Volence** [comment]
I think it would be very helpful at this point to have listed out the mechanics, actual numbers, and intended levels and tiering of the machine (or at least what's planned and what still needs to be implemented), and when the tiers will be available because I know the machine has changed scope a bit and it's a little difficult to get a good judgement of its balance and mechanics at this point

**Comment by OmdaCZ** [comment]
> I think it's a little awkward to have such a large multi for only 60 parallels - consider how many LPFs you could fit in that space, which are x3.5 speed (I'm not sure what the speed stats of this multi are). On checking, without even wallsharing them you can fit 39 to 40 LPFs in that footprint lengthwise, which seems to be a lot more throughput per space, while being both easier to integrate into non-lawnbase builds and easier to use in lines that just need a bit of fluid solidifying compared to the min size of this multi.
> 
> The cooling mechanic sounds cool, and the min-size footprint is workable, but the max-size footprint seems like the width-increasing-parallel mechanic was taken too far, esp. for a multi that comes in IV. This stands out _a lot_ from the other IV multis, and trying to make it scale from IV to UMV in such an integral mechanic makes it harder to potentially narrow that scaling later and introduce a late-game solidifer, compared to just making high-tier fluids stupid fast or having higher numbers to compensate for that late-game solidifer not existing yet.

What about 8+1/2*mWidth^2 ? would be 458 parallels max width with 4x speed compared to singleblock, while also discouraging spam of lower tier casing.

**Comment by Caedis** [comment]
was causing https://mclo.gs/9Q2tokI

**Comment by Dream-Master** [comment]
@OmdaCZ please add a recipe for the multi and the blocks in the core mod. Thanks

File: src/main/java/gregtech/common/tileentities/machines/multi/GT_MetaTileEntity_MultiSolidifier.java [hunk 1]
```diff
@@ -0,0 +1,678 @@
...
+        // Logic for GT_MetaTileEntity_Hatch_Solidifier
+        for (GT_MetaTileEntity_Hatch_Input solidifierHatch : mInputHatches) {
+            if (solidifierHatch instanceof GT_MetaTileEntity_Hatch_Solidifier) {
+                ItemStack mold = ((GT_MetaTileEntity_Hatch_Solidifier) solidifierHatch).getMold();
+                FluidStack fluid = solidifierHatch.getFluid();
+
+                if (mold != null && fluid != null) {
+                    List<ItemStack> inputItems = new ArrayList<>();
+                    inputItems.add(mold);
+                    inputItems.add(GT_Utility.getIntegratedCircuit(22));
```
Comment (NotAPenguin0) [inline src/main/java/gregtech/common/tileentities/machines/multi/GT_MetaTileEntity_MultiSolidifier.java:623]: This call should be removed, the solidifier multi does not need a circuit 22 to do recipes like the LPF does.

File: src/main/java/gregtech/common/tileentities/machines/multi/GT_MetaTileEntity_MultiSolidifier.java [hunk 2]
```diff
@@ -0,0 +1,678 @@
...
+        }
+        return tieredPipeCasings.get(meta)
+            .getRight();
+    }
+
+    private boolean addToSolidifierList(IGregTechTileEntity aTileEntity, int aBaseCasingIndex) {
+        if (aTileEntity == null) {
+            return false;
+        }
+        IMetaTileEntity aMetaTileEntity = aTileEntity.getMetaTileEntity();
```
Comment (NotAPenguin0) [inline src/main/java/gregtech/common/tileentities/machines/multi/GT_MetaTileEntity_MultiSolidifier.java:481]: This method definitely already exists in the superclass.

**Review by NotAPenguin0 — CHANGES_REQUESTED** [review]
Aside from some code issues, I'm very confused by the tiering of this multiblock. It seems to support various casing tiers, but I can't find any place in the code where they are actually used. If this is something you added and then removed, please remove the support for those casings too. 

I think the maximum width should be adjusted to be more in line with the size of the other LPF multiblocks. Currently the multi is absolutely huge but doesn't scale much further than the others. I also think the size upgrade is a bit odd, because there doesn't seem to be any cost to it other than a few relatively cheap casing blocks, but maybe that's fine for a multi that's not supposed to scale to stargate. I think some kind of scaling with energy tier or glass tier would be fine, maybe alongside the width scaling? Perhaps something like the following
- Glass tier determines energy hatch tier
- Energy hatch tier determines maximum width

**Comment by Dream-Master** [comment]
@OmdaCZ @FourIsTheNumber any news here ?

File: src/main/java/gregtech/common/tileentities/machines/multi/GT_MetaTileEntity_MultiSolidifier.java [hunk 3]
```diff
@@ -264,21 +268,28 @@ protected GT_Multiblock_Tooltip_Builder createTooltip() {
...
-            .addCasingInfoMin("Radiator casing/", 18, false)
+            .addCasingInfoMin("Tier 1: Solidifier Casing", 146, true)
+            .addCasingInfoMin("Tier 2: Laurenium Casing", 146, true)
+            .addCasingInfoMin("Tier 3: DTPF Casing", 146, true)
+            .addCasingInfoMin("Tier 1: Radiator Casing", 18, true)
+            .addCasingInfoMin("Tier 2: Cinobite Pipe Casing", 18, true)
+            .addCasingInfoMin("Tier 3: Abyssal Alloy Pipe Casing", 18, true)
+            .addCasingInfoMin("Heat Proof Casing", 4, false)
+            .addCasingInfoMin("Solid Steel Casing", 4, false)
+            .addInfo("Casings/Pipe Casings limit maximal width 2; 4; 6")
```
Comment (NotAPenguin0) [inline src/main/java/gregtech/common/tileentities/machines/multi/GT_MetaTileEntity_MultiSolidifier.java:292]: Just write `Tier limits...`

File: src/main/java/gregtech/common/tileentities/machines/multi/GT_MetaTileEntity_MultiSolidifier.java [hunk 4]
```diff
@@ -348,6 +359,15 @@ else return survivialBuildPiece(
 
     private int mCasingAmount;
 
+    int mTier; {
```
Comment (NotAPenguin0) [inline src/main/java/gregtech/common/tileentities/machines/multi/GT_MetaTileEntity_MultiSolidifier.java:362]: This needs to be updated on every structure check. You should also simplify the logic to `mTier = Math.max(pipeCasingTier, casingTier);`

File: src/main/java/gregtech/common/tileentities/machines/multi/GT_MetaTileEntity_MultiSolidifier.java [hunk 5]
```diff
@@ -315,7 +326,7 @@ public int survivalConstruct(ItemStack stackSize, int elementBudget, ISurvivalBu
         nWidth = 0;
         int built = survivialBuildPiece(STRUCTURE_PIECE_MAIN, stackSize, 3, 4, 0, elementBudget, env, false, true);
         if (built >= 0) return built;
-        int tTotalWidth = Math.min(6, stackSize.stackSize + 3);
+        int tTotalWidth = Math.min(3 + getCasingTier(), stackSize.stackSize + 3);
```
Comment (NotAPenguin0) [inline src/main/java/gregtech/common/tileentities/machines/multi/GT_MetaTileEntity_MultiSolidifier.java:329]: Shouldn't this be using `mTier`?

**Review by TheEpicGamer274 — APPROVED** [review]
Looks fine to me

**Pull request merged**